### PR TITLE
Alias Handler for SCIM Users

### PR DESCRIPTION
**Pull request by adrianhoelzl-sap**
see issue #2505 

Preparation for alias feature for SCIM users; adds subclass of `EntityAliasHandler` that handles alias entities of SCIM users.

Prerequisites:

- #2765
- #2824

### CODE REVIEW

File: server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java
Comment (Tallicia) [inline server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java:61]: This should be setPasswordLastModifiedTime to align to the naming of the two following Time setters.

File: server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java
Comment (Tallicia) [inline server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java:64]: Why are these setting the Time to the existingAliasEntity instead of "now" or even null for the 2 logon times -  a new Alias should have it's own timestamps. Setting them to the existingAlias seems it would be confusing when it was actually Modified or logged into.

File: server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java
Comment (strehle) [inline server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java]: the 2nd place for aliasEntitiesEnabled , maybe create an extra bean or component and use here a ref

File: server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java
Comment (strehle) [inline server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java:146]: did not found a test which cover this part

File: server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java
Comment (strehle) [inline server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java]: you dont need to set objects to null, why these many calls with null

File: server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java
Comment (strehle) [inline server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java:110]: empty is not what I would expect. is this a required field?

File: server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java
Comment (strehle) [inline server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java]: We do in other places Json object copy, e.g.

ScimUser scimCopy = JsonUtils.convertValue(originalEntity, ScimUser.class);

File: server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java
Comment (adrianhoelzl-sap) [reply server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java:110]: In the create user endpoint, we also set the password explicitly to an empty string for all users with a non-`uaa` origin: https://github.com/cloudfoundry/uaa/blob/6c30c8552fc9c9d4010d9b2c47d6b7771e42210a/server/src/main/java/org/cloudfoundry/identity/uaa/scim/endpoints/ScimUserEndpoints.java#L215

File: server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java
Comment (adrianhoelzl-sap) [reply server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java]: I wanted to make it explicit in this method which fields are set and which are `null`.

File: server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java
Comment (adrianhoelzl-sap) [reply server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java]: see the comment above

File: server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java
Comment (adrianhoelzl-sap) [reply server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java:146]: unit test added in commit e78e1fd

File: server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java
Comment (adrianhoelzl-sap) [reply server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java]: How should this bean look like? Just a wrapper for the boolean value?

**Comment by peterhaochen47** [comment]
Hi, as this PR is a stepping stone toward the overall "alias feature," and as it looks like the changes are limited to the specific code paths related to the feature, I would defer the approval of this PR to @strehle, in terms of judging 1) whether this PR accomplishes its goal, and 2) wether this PR has enough test coverage and user docs (if applicable). 

I'll also approve as well as long as our prior understanding the "alias feature" are still true after this PR: That the functionality & performance of all the existing UAA features are not impacted (do not experience breaking changes / regressions) when the "alias feature" is not turned on (and the feature is off by default).

File: server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java
Comment (Tallicia) [reply server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java]: Any update on this question?

File: server/src/test/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandlerValidationTest.java
Comment (Tallicia) [inline server/src/test/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandlerValidationTest.java]: extra linefeed to remve.

Good tests, nice modular implementation.

**Review by Tallicia — APPROVED** [review]
Approved with request to answer questions before merging.

File: server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java
Comment (strehle) [reply server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java]: I would like to see a configuration either as single option or a more detailed one. You do not need a complete configuration class like ratelimit, e.g. https://github.com/cloudfoundry/uaa/blob/develop/server/src/main/java/org/cloudfoundry/identity/uaa/ratelimiting/config/InitialConfig.java , but a central configuration in uaa-release should be created and documented and then you can have a single bean/component in uaa.

File: server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java
Comment (strehle) [reply server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java]: sounds not clear to me, because in java the variables are null after initialization, do you not trust JVM ?

File: server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java
Comment (strehle) [inline server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java:50]: This part is not covered, please add a Unit or IT

File: server/src/main/java/org/cloudfoundry/identity/uaa/alias/EntityAliasHandler.java
Comment (strehle) [inline server/src/main/java/org/cloudfoundry/identity/uaa/alias/EntityAliasHandler.java:228]: This part is not covered, please add a Unit or IT

**Review by strehle — COMMENTED** [review]
I would like to see 100 coverage on lines at least. Coverage on all conditions might be hard, but we should have tests with a meaning on all code lines.

Tests are like a technical specification, they later help to understand the code

File: server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java
Comment (adrianhoelzl-sap) [reply server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java:61]: The setters are already named consistently to the underlying fields in the class as well as its JSON representation (`passwordLastModified` does not end with "time", `lastLogonTime` and `previousLogonTime` do), see for example here: https://github.com/cloudfoundry/uaa/blob/091fbf2416472d9278286ce1aed3b41b5e22f8a5/model/src/main/java/org/cloudfoundry/identity/uaa/scim/impl/ScimUserJsonDeserializer.java#L92

I would therefore suggest to leave them as they are. What do you think?

File: server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java
Comment (adrianhoelzl-sap) [reply server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java:64]: This method is used for managing the fields that should differ between the "original" user and its alias, i.e., the properties of an alias that should be independent from the original user. 

During updates, we build a copy of the original user and leave these three timestamps empty. After that, we call this method to overwrite the timestamps with the values from the version of the alias prior to the update.

File: server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java
Comment (adrianhoelzl-sap) [reply server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java]: fixed with commit 5699f43

File: server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java
Comment (adrianhoelzl-sap) [reply server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java:50]: added test with commit 97d13b3

File: server/src/main/java/org/cloudfoundry/identity/uaa/alias/EntityAliasHandler.java
Comment (adrianhoelzl-sap) [reply server/src/main/java/org/cloudfoundry/identity/uaa/alias/EntityAliasHandler.java:228]: added test with commit 4321693

File: server/src/test/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandlerValidationTest.java
Comment (adrianhoelzl-sap) [reply server/src/test/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandlerValidationTest.java]: fixed with commit 31ebf13

File: server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java
Comment (adrianhoelzl-sap) [reply server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java]: introduced bean for flag value with commit 823fc2f

**Comment by strehle** [comment]
@adrianhoelzl-sap please cleanup PR , see 
https://sonarcloud.io/project/issues?resolved=false&sinceLeakPeriod=true&pullRequest=2769&id=cloudfoundry-identity-parent

Rest is ok

File: server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java
Comment (adrianhoelzl-sap) [reply server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java]: PR for specification of flag in uaa-release repo: https://github.com/cloudfoundry/uaa-release/pull/899

**Review by strehle — APPROVED** [review]

File: server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java
Comment (Tallicia) [reply server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java:61]: Seems the naming inconsistency is more widespread, maybe this would be a good clean up following this PR.

Comment (Tallicia) [reply server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java:61]: Seems the naming inconsistency is more widespread, maybe this would be a good clean up following this PR.

File: server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java
Comment (Tallicia) [reply server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java:64]: Which one is it, 

> "the properties of an alias that should be independent from the original user."

which I agree with
But this seems contrary to keeping them independent:

> "After that, we call this method to overwrite the timestamps with the values from the version of the alias prior to the update."

**Comment by Tallicia** [comment]
> @adrianhoelzl-sap please cleanup PR , see https://sonarcloud.io/project/issues?resolved=false&sinceLeakPeriod=true&pullRequest=2769&id=cloudfoundry-identity-parent
> 
> Rest is ok

@strehle This link is not showing no results for me: 

Screenshot 2024-06-04 at 9 11 40 AM

**Comment by strehle** [comment]
> > @adrianhoelzl-sap please cleanup PR , see https://sonarcloud.io/project/issues?resolved=false&sinceLeakPeriod=true&pullRequest=2769&id=cloudfoundry-identity-parent
> > Rest is ok
> 
> @strehle This link is not showing no results for me:
> 
> Screenshot 2024-06-04 at 9 11 40 AM

there where updates and now smells are solved
See overview https://sonarcloud.io/summary/new_code?id=cloudfoundry-identity-parent&pullRequest=2769

File: server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java
Comment (adrianhoelzl-sap) [reply server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java:64]: Sorry for the confusion in my explanation, I'll try again:

Let's say there is a user `U` in the "uaa" zone, which has an alias user `A` in the zone "custom". Whenever we perform an update on `U`, we persist the changes, which leads to a newer version `U'` with the changed properties.

Then, to propagate the changes also to the alias, we build a new clone of `U'`, i.e., `A'`. This is done here: https://github.com/cloudfoundry/uaa/blob/f451e8eac58919f28e07330ac04cce9687cb593b/server/src/main/java/org/cloudfoundry/identity/uaa/scim/ScimUserAliasHandler.java#L87

However, as you correctly addressed, `A'` would now have the same timestamp values (last logon, password last modified and previous logon) as `U'`, which is incorrect. Therefore, before persisting `A'`, we look up the version of `A'` before the update, i.e., `A`, and overwrite the timestamp values of `A'` with the timestamp values of `A`.

This is done in the method of this GitHub conversation. The parameter `newAliasEntity` corresponds to `A'`, while `existingAliasEntity` corresponds to `A` (and not to `U` or `U'`, as you might have thought when asking the question).

**Comment by strehle** [comment]
@Tallicia do you have open issues here ?

**Comment by torsten-sap** [comment]
I merge this PR as all questions seem to be addressed and two approvals exist.

**Pull request merged**
